documentation/API-Testing-Guidelines.md

```diff
@@ -1,41 +1,288 @@
+* For subscriptions that provide `subscriptionMaxEvents`, validate that the subscribed events are not longer received after the maximum events limit is reached..
+* Validate that after a subscription is deleted, the subscribed events are not longer received.
+
+### Error scenarios <a name="error-scenarios"></a>
+
+* All errors explicitly documented in the API spec must be covered by one or more dedicated error scenarios.
+  - In particular operations that expect a `device` object allowing one or more identifiers to be provided, must include scenarios to test all error cases defined. 
+
+* It must be validated that the HTTP status code and response property `code` are exactly those specified in the API spec.
+
+* It must be validated that the property `message` is present, but its value is free for the implementation to decide. Test plan may validate that message is human friendly and coherent with the scenario.
+
+For operations with a request body, several scenarios are expected to test different sets of input data which do not comply with the request body schema, testing that an error, typically `400` `INVALID_ARGUMENT` is received. Typical tests:
+- Values with invalid format
+- Mandatory properties are missing
+
+## Test plan design guidelines <a name="guidelines"></a>
 
 * Granularity of the feature file must be decided at the project level but it is recommended to:
     -	group in one file all scenarios testing one closely related API capability (that can cover one or several endpoints).
     -	provide several files when one CAMARA API (yaml) covers several independent functions that can be provided independently
 
-* Third Person pronoun usage in feature file is advisable as using the third person, conveys information in a more official manner.
-* It is recommended to only have one When/Then in a feature file as per cucumber official documentation's recommendations. However, in case of  complex scenarios, several when/then can be allowed.
-* The recommended format for scenario identifier is shown below
+* Third person pronoun usage in feature file is advisable as using the third person, conveys information in a more official manner.
+
+
+### Location of feature files <a name="location"></a>
+
+The feature files will reside under 
+
+`{Subproject_Repository}/code/Test_definitions/file_name.feature`
+
+E.g.: 
+
+`https://github.com/camaraproject/DeviceLocation/blob/main/code/Test_definitions/location-verification.feature`
+
+
+Filename:
+
+- APIs with a single feature file will use the api-name as file name: E.g. `location-verification.feature`
+- For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the `operationId` as file name: `createSession.feature`.
+- For APIs with several feature files, not split by operation, may use the api-name along some other description, e.g.: `qod-notifications.feature`
```

Comment:
@jlurien  We think API name should be mandatory with some added description if needed .

Reply:
Fine to me, I could suggest having always the api-name and, when the testing plan is composed of several files, that suffix, which can be operationId or other description.

Reply:
Suggestion from @rartych applied

Reply:
@jlurien  File still does not contain apiname-operationid/description in 2nd point  for Createsession.feature.
Kindly amend in the PR

Reply:
I think adding api-name in the second point brings consistency (filename always starts with api-name) , but can be helpful in the future with multiple API specifications in the same repository (distinguishing tests in case the same operationId is used).

> - For APIs with several operations that split the test plan in one feature file per operation, recommendation is to use the api-name along the `operationId` as file name: `device-identifier-retrieveType.feature`.

Reply:
I agree. Applied